File: 1905.统计子岛屿.go
/*
 * @lc app=leetcode.cn id=1905 lang=golang
 *
 * [1905] 统计子岛屿
 */

// @lc code=start
func countSubIslands(grid1 [][]int, grid2 [][]int) (res int) {
	m, n := len(grid1), len(grid1[0])
	directions := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	var dfs func(x, y int)
	dfs = func(x, y int) {
		// 内部递归时只处理陆地格子，不能越界
		if !(x < m && x >= 0 && y < n && y >= 0 && grid2[x][y] == 1) {
			return
		}
		grid2[x][y] = 0 // 剪枝

		for _, direction := range directions {
			dfs(x+direction[0], y+direction[1])
		}
	}
	// 清除不符合条件的岛屿
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid1[i][j] == 0 && grid2[i][j] == 1 {
				dfs(i, j)
			}
		}
	}
	// 计算子岛屿
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid2[i][j] == 1 {
				res++
				dfs(i, j)
			}
		}
	}
	return
}

// @lc code=end

File: 1905.统计子岛屿.py
#
# @lc app=leetcode.cn id=1905 lang=python3
#
# [1905] 统计子岛屿
#

# @lc code=start
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        # 得到矩阵的行和列
        m, n = len(grid1), len(grid1[0])
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        # dfs，搜索一片岛屿
        def dfs(x, y):
            if not (0 <= x < m and 0 <= y < n and grid2[x][y] == 1):
                return
            # 将搜索过的岛屿置0，防止重复搜索
            grid2[x][y] = 0

            for direction in directions:
                dfs(x + direction[0], y + direction[1])

        # 首先将 2 中岛屿不被 1 中包含的剔除掉
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1 and grid1[i][j] == 0:
                    dfs(i, j)

        # 计算子岛屿数量
        res = 0
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1:
                    res += 1
                    dfs(i, j)
        return res
# @lc code=end
